lib/cache/inventory_cache.go

```diff
@@ -0,0 +1,678 @@
+
+	// subLinearComponent ensures that growth is sub-linear across most reasonable cluster sizes.
+	subLinearComponent := int(math.Sqrt(float64(clusterSize)))
+
+	return minimumComponent + linearComponent + subLinearComponent
+}
+
+// initializeAndWatchWithRetry runs initializeAndWatch with a retry every 10 seconds if it fails.
+func (ic *InventoryCache) initializeAndWatchWithRetry(ctx context.Context) {
+	defer close(ic.initDone)
+
+	const retryInterval = 10 * time.Second
+
+	for {
+		ic.cfg.Logger.DebugContext(ctx, "Attempting to initialize inventory cache")
+
+		// Attempt to initialize and watch
+		if err := ic.initializeAndWatch(ctx); err != nil {
+			if ctx.Err() != nil {
+				return
+			}
+
+			ic.cfg.Logger.WarnContext(ctx, "Failed to initialize inventory cache, retrying in 10 seconds",
+				"error", err)
+
+			// Wait before retrying
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(retryInterval):
+			}
+			continue
+		}
+
+		return
+	}
+}
+
+// initializeAndWatch initializes the inventory cache and begins watching for instance and bot_instance backend events.
+func (ic *InventoryCache) initializeAndWatch(ctx context.Context) error {
+	// Wait for primary cache to be ready.
+	if err := ic.waitForPrimaryCacheInit(ctx); err != nil {
+		return trace.Wrap(err, "Failed to wait for primary cache init")
+	}
+
+	// Setup the backend watcher.
+	watcher, err := ic.setupWatcher(ctx)
+	if err != nil {
+		return trace.Wrap(err, "Failed to set up backend watcher")
+	}
+	defer watcher.Close()
+
+	// Wait for the watcher to be ready.
+	if err := ic.waitForWatcherInit(ctx, watcher); err != nil {
+		return trace.Wrap(err, "Failed to wait for watcher init")
+	}
+
+	// Calculate the rate limit to use.
+	primaryCacheSize := ic.getPrimaryCacheSize()
+	readsPerSecond := calculateReadsPerSecond(primaryCacheSize)
+
+	// Populate the cache with teleport instance and bot instances.
+	if err := ic.populateCache(ctx, readsPerSecond); err != nil {
+		return trace.Wrap(err, "failed to populate inventory cache")
+	}
+
+	// Mark cache as healthy.
+	ic.healthy.Store(true)
+
+	// This runs infinitely until the context is cancelled, so the return below won't be hit until shutdown.
+	ic.processEvents(ctx, watcher)
+
+	return nil
+}
+
+// waitForPrimaryCacheInit waits for the primary cache to be initialized.
+func (ic *InventoryCache) waitForPrimaryCacheInit(ctx context.Context) error {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+
+		case <-ticker.C:
+			ic.cfg.PrimaryCache.rw.RLock()
+			ok := ic.cfg.PrimaryCache.ok
+			ic.cfg.PrimaryCache.rw.RUnlock()
+
+			if ok {
+				return nil
+			}
+		}
+	}
+}
+
+// setupWatcher sets up a backend watcher for instance and bot_instance events.
+func (ic *InventoryCache) setupWatcher(ctx context.Context) (backend.Watcher, error) {
+	watcher, err := ic.cfg.Backend.NewWatcher(ctx, backend.Watch{
```

Comment:
Create a `local.EventsService` on top of the backend (or maybe just take a `services.Events` rather than a `backend.Backend`, the auth definitely has one), don't reimplement the parsers for `instance` and `bot_instance` here.